### PR TITLE
Support runtime chunk deduplication

### PR DESCRIPTION
**Pull request by jiangliu**
## Details
This PR enhances `nydusd` to support runtime chunk deduplication. It works in this way:
1) Use a sqlite database to record information about decompressed/plaintext chunks available on local node.
2) When a chunk is not ready in the uncompressed data blob file, query the sqlite database whether a chunk with the same chunk digest is available. If a chunk with the same chunk digest exists, copy the decompressed from the source data blob file to the target data blob by using `copy_file_range()`.
3) Otherwise download the compressed chunk from remote, uncompress it and write to the target data blob, and add a record for the chunk to the database.

So there are two types of chunk deduplication:
1) saving network bandwidth when the chunk is available on local node, because we don't need to download compressed chunk data from remote.
2) saving local disk space if the underlying filesystem supports reference. If the filesystem storing data blob files supports reference, `copy_file_range()` will optimize to use reference instead of data copy, thus reduce local storage consuption.  

## Types of changes

_What types of changes does your PullRequest introduce? Put an `x` in all the boxes that apply:_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Go over all the following points, and put an `x` in all the boxes that apply._

- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: smoke/tests/chunk_dedup.go
Comment (Desiki-high) [inline smoke/tests/chunk_dedup.go:1]: See https://go.dev/doc/tutorial/add-a-test, go test needs a file name like `xxxx_test.go`. How about renaming it to `chunk_dedup_test.go`.

Comment (jiangliu) [reply smoke/tests/chunk_dedup.go:1]: Thanks for reminder, fixed.

File: storage/src/cache/filecache/mod.rs [hunk 1]
```diff
@@ -266,14 +272,15 @@ impl FileCacheEntry {
                 );
                 return Err(einval!(msg));
             }
+            let load_chunk_digest = need_validation || cas_mgr.is_some();
```
Comment (ccx1024cc) [inline storage/src/cache/filecache/mod.rs:275]: If ( open chunk deduplication ) = ( no validation ), better to stat in doc too.

**Review by ccx1024cc — COMMENTED** [review]
NICE. LGTM.
`copy_file_range` is triggered in tests. It's effective.

File: storage/src/cache/fscache/mod.rs [hunk 2]
```diff
@@ -266,13 +273,20 @@ impl FileCacheEntry {
         )?));
         Self::restore_chunk_map(blob_info.clone(), file.clone(), &meta, &chunk_map);
 
+        let mut blob_data_file_path = String::new();
+        if cas_mgr.is_some() {
+            blob_data_file_path = get_path_from_file(&file);
```
Comment (ccx1024cc) [inline storage/src/cache/fscache/mod.rs:278]: `get_path_from_file` is not recovered in tests. `cas_mgr` is empty?

[coverage(https://app.codecov.io/gh/dragonflyoss/nydus/pull/1507/blob/storage/src/utils.rs)

**Review by ccx1024cc — CHANGES_REQUESTED** [review]

**Comment by ccx1024cc** [comment]
> NICE. LGTM. `copy_file_range` is triggered in tests. It's effective.

Sorry, not LGTM. Doubt that copy_file_range is covered by ut, not by smoke case.
`get_path_from_file` is key path for fileEntry with cas manager. If smoke case works, then coverage is expected.

**Comment by ccx1024cc** [comment]
> > NICE. LGTM. `copy_file_range` is triggered in tests. It's effective.
> 
> Sorry, not LGTM. Doubt that copy_file_range is covered by ut, not by smoke case. `get_path_from_file` is key path for fileEntry with cas manager. If smoke case works, then coverage is expected.

That is point that new feature not opened in smoke.
[Args coverage](https://app.codecov.io/gh/dragonflyoss/nydus/pull/1507?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=dragonflyoss#diff-c3RvcmFnZS9zcmMvY2FjaGUvZGVkdXAvZGIucnM=)

File: smoke/tests/chunk_dedup_test.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+		{
+			Digest:         lowerRafsBlobDigest,
+			OriginalDigest: &lowerOCIBlobDigest,
+		},
+	})
+	require.Equal(t, []digest.Digest{lowerOCIBlobDigest}, actualDigests)
+
+	// Verify lower layer mounted by nydusd
+	ctx.Env.BootstrapPath = lowerBootstrap
+	tool.Verify(t, ctx, lowerLayer.FileTree)
```
Comment (imeoer) [inline smoke/tests/chunk_dedup_test.go:70]: We may need a way to check if the CAS works.

File: storage/src/cache/fscache/mod.rs [hunk 4]
```diff
@@ -240,9 +240,16 @@ impl FileCacheEntry {
         };
         let blob_compressed_size = Self::get_blob_size(&reader, &blob_info)?;
 
+        // Turn off chunk deduplication in case of tarfs.
+        let cas_mgr = if is_tarfs {
```
Comment (imeoer) [inline storage/src/cache/fscache/mod.rs:244]: Maybe give a warnning message about conflict feature here.

File: storage/src/cache/filecache/mod.rs [hunk 5]
```diff
@@ -208,11 +209,18 @@ impl FileCacheEntry {
         } else {
             reader.clone()
         };
+        // Turn off chunk deduplication in case of cache data encryption is enabled or is tarfs.
+        let cas_mgr = if mgr.cache_encrypted || mgr.cache_raw_data || is_tarfs {
```
Comment (imeoer) [inline storage/src/cache/filecache/mod.rs:213]: Maybe give a warn message about conflict feature here.

File: docs/data-deduplication.md [hunk 6]
```diff
@@ -164,4 +164,25 @@ So Nydus provides a node level CAS system to reduce data downloaded from the reg
...
 The node level CAS system helps to achieve O4 and O5.
 
-# Node Level CAS System (WIP)
+# Node Level CAS System
+Data deduplication can also be achieved when accessing Nydus images. The key idea is to maintain information about data chunks available on local host by using a database.
+When a chunk is needed but not available in the uncompressed data blob files yet, we will query the database using chunk digest as key. 
+If a record with the same chunk digest already exists, it will be reused.
+We call such a system as CAS (Content Addressable Storage).
+
+## Chunk Deduplication by Using CAS as L2 Cache
```
Comment (imeoer) [inline docs/data-deduplication.md:173]: This still seems to be an experimental feature, do we still need to consider the `cas.db` record recycling?

**Comment by jonoirwinrsa** [comment]
Hi all, I tried out this feature and it seems to work as expected. Is there something preventing it from being merged?

**Comment by imeoer** [comment]
> Hi all, I tried out this feature and it seems to work as expected. Is there something preventing it from being merged?

cc @jiangliu any updates we can continue? :)

**Comment by elijah-rou** [comment]
what is the status of this PR? anything to do to help get it merged? :)

**Comment by Desiki-high** [comment]
Complete in https://github.com/dragonflyoss/nydus/pull/1626.

**Pull request closed without merging**